File: solutions/0338-counting-bits/solution.java
class Solution {
    public int[] countBits(int n) {
        int[] ans = new int[n + 1];

        for (int i = 0; i <= n; i++) {
            String bit = decimalToBinary(i);
            int count = 0;
            
            for (int j = 0; j < bit.length(); j++) {
                if (bit.charAt(j) == '1') {
                    count++;
                }
            }
            ans[i] = count;
        }
        return ans;
    }

    private String decimalToBinary(int num) {
        if (num == 0) return "0"; // Handle edge case explicitly

        StringBuilder binary = new StringBuilder();
        while (num > 0) {
            binary.insert(0, num % 2); // Prepend binary digits
            num /= 2;
        }
        return binary.toString();
    }
}

File: solutions/0338-counting-bits/solution.js
/**
 * @param {number} n
 * @return {number[]}
 */

var countBits = function(n) {
function decimalToBinary(num) {
        let binary = "";
        while (num > 0) {
            binary = (num % 2) + binary; // Append remainder (0 or 1)
            num = Math.floor(num / 2); // Divide by 2
        }
        return binary === "" ? "0" : binary; // Handle case when num = 0
    }
    let ans =[];
    for(let i=0;i<=n;i++){
        let count=0;
        let bit=decimalToBinary(i);
        let bitArr= bit.split('')
        for(let j=0;j<bitArr.length;j++){
            if(bitArr[j]==1){
                count += 1;
            }
        }
        ans[i] = count;
    }
    return ans;
};
